### PR TITLE
feat: extend plugin API supports communication

### PR DESCRIPTION
**Pull request by airslice**
close https://github.com/reearth/reearth/issues/376

# Overview

This PR enables plugin communication with other plugins.

## What I've done

Implement the API:

```ts
type PluginExtensionInstance = {
  id: string;
  pluginId: string;
  name: string; 
  extensionId: string;
  extensionType: "widget" | "block";
};

reearth.plugins.instances: PluginExtensionInstance[]
reearth.plugins.postMessage(id: string; message: any)
reearth.on("pluginmessage", (e: { data: any, sender: string }) => {})
```

## How I tested

Test with plugin:
[plugincomm.zip](https://github.com/reearth/reearth-web/files/10172887/plugincomm.zip)

Download & install & add multiple instances for test.
Also add block for test.

## Screenshot
localhost_3000_edit_01gjy7yqpqnjzcvj0wsk7h0f4w_preview(1440 desktop)

### CODE REVIEW

File: src/components/molecules/Visualizer/hooks.ts [hunk 1]
```diff
@@ -257,6 +260,12 @@ export default ({
     alignSystem,
   });
 
+  const { pluginInstances } = usePluginInstances({
+    alignSystem,
+    floatingWidgets,
+    blocks: infobox?.blocks ?? selectedLayer?.infobox?.blocks,
```
Comment (KaWaite) [inline src/components/molecules/Visualizer/hooks.ts:266]: Doesn't `infobox` get assigned from `selectedLayer?.infobox` in `useLayers`? If that is correct, we don't need the fallback because they are the same thing

File: src/components/molecules/Visualizer/usePluginInstances.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  }, []);
+
+  const pluginInstances = useMemo(() => {
+    return {
+      meta: pluginInstancesMeta,
+      postMessage: pluginPostMessage,
+      addPluginMessageSender,
+      removePluginMessageSender,
+    };
+  }, [pluginInstancesMeta, pluginPostMessage, addPluginMessageSender, removePluginMessageSender]);
```
Comment (KaWaite) [inline src/components/molecules/Visualizer/usePluginInstances.ts:103]: Why is this necessary instead of just returning them normally? I feel like this isn't necessary but guessing you did this for a specific reason.
But, all of the returned values are either useState or useCallback, which if they change should force a re-render.

File: src/components/molecules/Visualizer/hooks.ts (hunk 1 above)
Comment (airslice) [reply src/components/molecules/Visualizer/hooks.ts:266]: Yes you're right. Should not use `infobox?.blocks` here. Updated.

**Review by rot1024 — CHANGES_REQUESTED** [review]

File: src/components/molecules/Visualizer/usePluginInstances.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+  const pluginMessageSenders = useRef<Map<string, (msg: any) => void>>(new Map());
+
+  const pluginInstances = useMemo(() => {
+    return {
+      meta: pluginInstancesMeta,
+      postMessage: (id: string, msg: any, sender: string) => {
+        const msgSender = pluginMessageSenders.current?.get(id);
+        if (!msgSender) return;
+        msgSender({ data: msg, sender });
```
[the remark this answers is not in the document]
Comment (airslice) [reply src/components/molecules/Visualizer/usePluginInstances.ts]: Updated.

**Review by rot1024 — APPROVED** [review]
